README.md
# Energy Monitoring Interface

EnergyMon provides a general C interface for energy monitoring utilities.

If using this project for other scientific works or publications, please reference:

* Connor Imes, Lars Bergstrom, and Henry Hoffmann. "A Portable Interface for Runtime Energy Monitoring". In: FSE. 2016. DOI: https://doi.org/10.1145/2950290.2983956
  <details>
  <summary>[BibTex]</summary>

  ```BibTex
  @inproceedings{imes2016energymon,
    author = {Imes, Connor and Bergstrom, Lars and Hoffmann, Henry},
    title = {A Portable Interface for Runtime Energy Monitoring},
    year = {2016},
    isbn = {9781450342186},
    publisher = {Association for Computing Machinery},
    address = {New York, NY, USA},
    url = {https://doi.org/10.1145/2950290.2983956},
    doi = {10.1145/2950290.2983956},
    booktitle = {Proceedings of the 2016 24th ACM SIGSOFT International Symposium on Foundations of Software Engineering},
    pages = {968–974},
    numpages = {7},
    keywords = {portable energy measurement},
    location = {Seattle, WA, USA},
    series = {FSE 2016}
  }
  ```

  </details>
* You may also find an extended analysis in the [Tech Report](https://cs.uchicago.edu/research/publications/techreports/TR-2016-08).
  <details>
  <summary>[BibTex]</summary>

  ```BibTex
  @techreport{imes2016energymon-tr,
    author = {Imes, Connor and Bergstrom, Lars and Hoffmann, Henry},
    title = {A Portable Interface for Runtime Energy Monitoring: Extended Analysis},
    institution = {University of Chicago, Department of Computer Science},
    number = {TR-2016-08},
    year = {2016},
    month = sep,
    address = {Chicago, IL, USA}
  }
  ```

  </details>

Applications using some libraries may need to be executed using elevated privileges.

The following instructions are for Linux systems.
If you are using a different platform, change the commands accordingly.

Current EnergyMon implementation options are:

* **dummy** [default]: Mock implementation
* **cray-pm**: Cray XC30 and XC40 systems (e.g., NERSC Cori) via Linux sysfs files
* **ibmpowernv**: IBM PowerNV systems (e.g., OLCF Summit) via Linux sysfs energy sensor files
* **ibmpowernv-power**: IBM PowerNV systems (e.g., OLCF Summit) via Linux sysfs power sensor files
* **ipg**: Intel RAPL via `Intel Power Gadget`
* **jetson**: NVIDIA Jetson systems with INA3221 power sensors via Linux sysfs files (JetPack 4.x)
* **jetson-sensors**: NVIDIA Jetson systems with INA3221 power sensors via Linux sysfs files (JetPack 5.x)
* **msr**: Intel RAPL via Linux Model-Specific Register device files (supports most non-Atom CPUs)
* **odroid**: Hardkernel ODROID XU+E and XU3 systems (with INA-231 power sensors) via Linux sysfs files
* **odroid-ioctl**: Hardkernel ODROID XU+E and XU3 systems (with INA-231 power sensors) via `ioctl` on Linux device files
* **osp**: Hardkernel ODROID Smart Power meters (coarse-grained energy counter) via `HIDAPI`
* **osp-polling**: Hardkernel ODROID Smart Power meters (finer-grained power sensor) via `HIDAPI`
* **rapl**: Intel RAPL via Linux powercap sysfs files
* **raplcap-msr**: Intel RAPL via `libraplcap-msr` (more capable than `msr` implementation above)
* **shmem**: Shared memory client via an EnergyMon shared memory provider
* **wattsup**: Watts Up Pro meter via Linux device files
* **wattsup-libusb**: Watts Up Pro meter via `libusb`
* **wattsup-libftdi**: Watts Up Pro meter via `libftdi`
* **zcu102**: Xilinx Zynq UltraScale+ ZCU102 systems (with INA-226 power sensors) via Linux sysfs files

See the README files in subdirectories for implementation specifics, including dependencies.


## Building

This project uses CMake.

By default, all libraries will be built, with `dummy` as the `energymon-default` implementation:

``` sh
mkdir _build
cd _build
cmake ..
make
```

To use a different default implementation, e.g., the RAPL energy monitor, specify `ENERGYMON_BUILD_DEFAULT` with cmake:

``` sh
cmake -DENERGYMON_BUILD_DEFAULT=rapl ..
```

Set `ENERGYMON_BUILD_DEFAULT=NONE` to disable building a default implementation.
Its default value is `dummy`.

To build only a single library, e.g., the RAPL energy monitor, specify `ENERGYMON_BUILD_LIB` with cmake:

``` sh
cmake -DENERGYMON_BUILD_LIB=rapl ..
```

Set `ENERGYMON_BUILD_LIB=NONE` to only build the default implementation (if set).
Set back to its default value of `ALL` to build all libraries.

To build shared objects / dynamically linked libraries instead of static libraries, set `BUILD_SHARED_LIBS` with cmake:

``` sh
cmake .. -DBUILD_SHARED_LIBS=ON
```

For an optimized build, set `CMAKE_BUILD_TYPE` when with cmake, e.g., one of:

``` sh
cmake .. -DCMAKE_BUILD_TYPE=Release
cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
```

Of course, you can specify multiple options, e.g.:

``` sh
cmake .. -DENERGYMON_BUILD_LIB=NONE -DENERGYMON_BUILD_DEFAULT=rapl -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=Release
```

### Other Build Options

Boolean options:

 * `ENERGYMON_BUILD_SHMEM_PROVIDERS` - enable/disable building shared memory providers (True by default)
 * `ENERGYMON_BUILD_UTILITIES` - enable/disable building utility applications (True by default)
 * `ENERGYMON_BUILD_TESTS` - enable/disable building test code (True by default)
 * `ENERGYMON_BUILD_EXAMPLES` - enable/disable building examples (True by default)


## Installing

To install libraries, headers, and binaries, run with proper privileges:

``` sh
make install
```

On Linux, the installation usually places
libraries in `/usr/local/lib`,
header files in `/usr/local/include/energymon`, and
binary files in `/usr/local/bin`.


## Uninstalling

To remove files installed to the system, run with proper privileges:

``` sh
make uninstall
```


## Linking

### CMake

Projects that use `CMake` can find the `EnergyMon` package and link against imported targets, which automatically applies properties like header search paths and transitive dependencies:

```cmake
find_package(EnergyMon REQUIRED)

add_executable(hello_world hello_world.c)
target_link_libraries(hello_world PRIVATE EnergyMon::energymon-default)
```

### pkg-config

A more general approach for linking with an EnergyMon library is to use [pkg-config](http://www.freedesktop.org/wiki/Software/pkg-config/):

```sh
cc $(pkg-config energymon-default --cflags) hello_world.c \
   $(pkg-config energymon-default --libs --static) -o hello_world
```

or in a Makefile:

```Makefile
CFLAGS += $(shell pkg-config --cflags energymon-default)
LDFLAGS += $(shell pkg-config --libs --static energymon-default)

hello_world: hello_world.c
	$(CC) $(CFLAGS) hello_world.c -o $@ $(LDFLAGS)
```

If shared object libraries are installed, don't include the `--static` option.

Projects that use `CMake >= 3.6` can alternatively use a pkg-config `IMPORTED_TARGET`, rather than the direct CMake imports documented above:

```cmake
find_package(PkgConfig REQUIRED)
pkg_check_modules(EnergyMonDefault REQUIRED IMPORTED_TARGET energymon-default)

add_executable(hello_world hello_world.c)
target_link_libraries(hello_world PRIVATE PkgConfig::EnergyMonDefault)
```


## Usage

To use an EnergyMon implementation, you must first populate the struct by calling the getter function, then initialize it.
Don't forget to cleanup the instance once you're finished with it.
See `energymon.h` and `energymon-default.h` for more detailed function descriptions.

```C
  energymon em;
  uint64_t start_uj, end_uj;

  // get the energymon instance and initialize
  energymon_get_default(&em);
  em.finit(&em);

  // profile application function
  start_uj = em.fread(&em);
  do_work();
  end_uj = em.fread(&em);
  printf("Total energy for do_work() in microjoules: %"PRIu64"\n", end_uj - start_uj);

  // destroy the instance
  em.ffinish(&em);
```


## Tools

This project includes a handful of applications.

### Utilities

All of the following are linked with `energymon-default`:

* `energymon-cmd-profile`: Prints out time, energy, and power statistics for the execution of a given shell command.
* `energymon-power-poller`: Prints average power values at the requested interval for the previous interval period.
* `energymon-file-provider`: Writes energy data to a file (overwrites previous values).
* `energymon-idle-power`: Prints the average power over the given interval (meant to run in isolation and measure idle power consumption).
* `energymon-info`: Prints information about the implementation.
* `energymon-overhead`: Prints the latency overhead in nanoseconds of the functions `finit`, `fread`, and `ffinish`.

Implementation-specific versions of these utilities are also provided.

### Shared Memory Providers

Shared memory providers allow exposing energy data to one or more unprivileged applications for sources that require elevated privileges and/or exclusive access.
The providers may need to run with elevated privileges, but other applications can attach to their shared memory and read energy data using the `shmem` EnergyMon implementation.

* `energymon-osp-polling-shmem-provider`
* `energymon-wattsup-shmem-provider`
* `energymon-wattsup-libftdi-provider`
* `energymon-wattsup-libusb-provider`


## Project Source

Find this and related project sources at the [energymon organization on GitHub](https://github.com/energymon).  
This project originates at: https://github.com/energymon/energymon


Bug reports and pull requests for new implementations, bug fixes, and enhancements are welcome.


RELEASES.md
# Release Notes

## [Unreleased]

### Added

* jetson-sensors: new implementation for NVIDIA Jetson systems with JetPack 5.x

### Fixed

* jetson: some root cause errors like EACCES (Permission denied) are masked as ENODEV (No such device)
* pkg-config file is broken when CMAKE_INSTALL_{INCLUDE,LIB}DIR is absolute


## [v0.5.0] - 2022-05-13

### Added

* ibmpowernv: new implementation for IBM PowerNV systems (targeted for OLCF Summit)
* ipg: new implementation using the Intel Power Gadget 3.7 API (PowerGadgetLib.h)
* jetson: new implementation for NVIDIA Jetson systems
* energymon-file-provider: -F/--force option to force updates faster than the EnergyMon implementation claims
* energymon-info: -h/--help option to print help message
* energymon-overhead: -h/--help option to print help message
* energymon-power-poller: -F/--force option to force updates faster than the EnergyMon implementation claims
* shmem: man pages for shared memory providers

### Changed

* Internal tweaks based on testing with additional compiler flags

### Fixed

* odroid, odroid-ioctl, zcu102: save errno when cleaning up after init failure
* osp: link with librt on older systems
* shmem: enforce provider ID in range `[1, 255]`
* wattsup-libftdi: compilation fails with clang when `-Werror` option is used
* utils: use implementation-specific binary names in help messages (don't assume energymon-default name format)


## [v0.4.0] - 2021-05-12

### Added

* ibmpowernv-power: new implementation for IBM PowerNV systems (targeted for OLCF Summit)
* raplcap-msr: support for multi-die packages
* Travis CI: test on additional architectures

### Changed

* CMake: overhaul the project build system
  * Increased minimum CMake version from 2.8.5 to 3.6
  * Refactor library and binary target creation
  * Export CMake targets during install
* rapl: document Linux kernel changes/differences in README
* raplcap-msr: update minimum dependency version to 0.5.0
* wattsup-libftdi [#47]: prefer `ftdi_tcioflush` if available (`ftdi_usb_purge_buffers` deprecated in libftdi 1.5)
* energymon-idle-power: now accepts idle time in fractional seconds

### Fixed

* Valid 0 return values from fread functions mistaken for errors when errno is already set
* cray-pm: building as default libraries used wrong macro, causing build failures
* osp: some cumulative energy unaccounted for during overflow detection (precision loss)
* raplcap-msr: Requires.private field in pkg-config file should specify the minimum version for raplcap-msr dependency
* wattsup-libftdi: Description field in pkg-config file was empty
* energymon-power-poller: incorrect Pmin summary output when number of samples is 0


## [v0.3.1] - 2020-10-27

### Fixed

* rapl [#46]: Check that top-level zones are actually package (not psys) zones
* rapl: Set errno when when no zones are found during init


## [v0.3.0] - 2020-03-10

### Added

* raplcap-msr: New implementation using libraplcap-msr (more features than `msr` and `rapl` implementations)
* utils [#44]: Build utilities for all implementations, not just energymon-default
* test: Build tests for all implementations, not just energymon-default

### Changed

* man pages: More pedantic formatting
* RELEASES: Reformat file

### Fixed

* cmake [#14]: Print useful CMake error when bad ENERGYMON_BUILD_DEFAULT value is set
* rapl: Powercap directory paths use lower case hex values, not decimal values
* zcu102: Potential memory leak in 'get_sensor_directories' under failure conditions


## [v0.2.4] - 2018-06-01

### Changed

* WattsUp implementations:
  * Wait up to 5 seconds for a good data packet (or fail) before completing initialization
  * Purge data buffers during initialization
  * Poll device at shorter intervals to increase energy value accuracy
  * Better handling of incomplete data packets
* osp-polling: Allow polling thread to be interrupted during sleep when finish is requested
* energymon-power-poller: Account for true elapsed time to produce more accurate power values

### Fixed

* wattsup-libftdi, wattsup-libusb: Fix empty reads by always setting baud rate and serial line properties
* wattsup-dev: Use pselect to avoid timer expired errors
* odroid, odroid-ioctl, osp-polling, zcu102: Minor correction to polling delay interval
* msr: Possible (but unlikely) precision computation bug
* osp, osp-polling: Fix status check during init that didn't always work


## [v0.2.3] - 2017-11-19

### Added

* Implementation for Xilinx ZCU102 (Jason Miller)

### Fixed

* Fixed [#40]: Time conversion bug in common function 'energymon_gettime_elapsed_us' (Jason Miller)


## [v0.2.2] - 2017-10-26

### Added

* New capabilities to some utility applications:
  * energymon-cmd-profile: Accept commands without quotations; add help option
  * energymon-file-provider: Add help, count, interval, and no-rewind options
  * energymon-idle-power: Add help option
* Man pages for utility applications
* VERSION and SOVERSION properties on shared object libraries
* Multiarch support (use GNU standard installation directories)

### Changed

* Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
* Refactor this RELEASES.md file

### Removed

* Removed private symbol exports in shared object libraries (prevents symbol name clashes)

### Fixed

* Various bug fixes to utility applications


## [v0.2.1] - 2017-10-01

### Added

* Support for msr-safe kernel module in msr implementation
* New cray-pm implementations to support capabilities of Cray XC30 and XC40 systems
* New wattsup-libusb and wattsup-libftdi implementations
* AppVeyor test configuration for Windows continuous integration

### Changed

* Faster refresh in osp-polling, odroid and odroid-ioctl implementations
* Better multi-platform support for some implementations, especially with portable time and sleep abstractions
* Downgraded required CMake version from 2.8.9 to 2.8
* Documented FSE 2016 and Technical Report publication information

jetson-sensors/energymon-jetson-sensors.c
/**
 * Energy reading for NVIDIA Jetson devices with TI INA3221 power sensors.
 *
 * @author Connor Imes
 * @date 2022-08-04
 */
#include <assert.h>
#include <errno.h>
#include <inttypes.h>
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

// libsensors
#include <sensors.h>
#include <error.h>

#include "energymon.h"
#include "energymon-jetson-sensors.h"
#include "energymon-time-util.h"
#include "energymon-util.h"

#ifdef ENERGYMON_DEFAULT
#include "energymon-default.h"
int energymon_get_default(energymon* em) {
  return energymon_get_jetson_sensors(em);
}
#endif



// Environment variable to not call libsensors static lifecycle functions.
// Keeping this as an undocumented feature for now.
#define ENERGYMON_JETSON_SENSORS_SKIP_LIFECYCLE "ENERGYMON_JETSON_SENSORS_SKIP_LIFECYCLE"

// Environment variable to use a non-NULL sensors init file.
// Keeping this as an undocumented feature for now.
#define ENERGYMON_JETSON_SENSORS_INIT_FILE "ENERGYMON_JETSON_SENSORS_INIT_FILE"


#define INA3221_CHAN_NAME_LEN 64

#define JETSON_DEV_NAME_LEN 64
#define JETSON_INA3221_DEVS_MAX 2

typedef struct ina3221_chan_cfg {
  size_t id;
  char name[INA3221_CHAN_NAME_LEN];
} ina3221_chan_cfg;

#define INA3221_CHAN_CFG_INIT(_id, _name) { \
  .id = _id, \
  .name = _name }

typedef struct ina3221_dev_cfg {
  int bus;
  int addr;
  ina3221_chan_cfg channels[3];
} ina3221_dev_cfg;

static ina3221_dev_cfg DEVS_XAVIER_NX[JETSON_INA3221_DEVS_MAX] = {
  {
    .bus = 7,
    .addr = 0x40,
    .channels = {
      INA3221_CHAN_CFG_INIT(1, "VDD_IN"),
      INA3221_CHAN_CFG_INIT(2, "VDD_CPU_GPU_CV"),
      INA3221_CHAN_CFG_INIT(3, "VDD_SOC")
    },
  },
};

static ina3221_dev_cfg DEVS_AGX_XAVIER[JETSON_INA3221_DEVS_MAX] = {
  {
    .bus = 1,
    .addr = 0x40,
    .channels = {
      INA3221_CHAN_CFG_INIT(1, "GPU"),
      INA3221_CHAN_CFG_INIT(2, "CPU"),
      INA3221_CHAN_CFG_INIT(3, "SOC")
    },
  },
  {
    .bus = 1,
    .addr = 0x41,
    .channels = {
      INA3221_CHAN_CFG_INIT(1, "CV"),
      INA3221_CHAN_CFG_INIT(2, "VDDRQ"),
      INA3221_CHAN_CFG_INIT(3, "SYS5V")
    },
  },
};

static ina3221_dev_cfg DEVS_AGX_ORIN[JETSON_INA3221_DEVS_MAX] = {
  {
    .bus = 1,
    .addr = 0x40,
    .channels = {
      INA3221_CHAN_CFG_INIT(1, "VDD_GPU_SOC"),
      INA3221_CHAN_CFG_INIT(2, "VDD_CPU_CV"),
      INA3221_CHAN_CFG_INIT(3, "VIN_SYS_5V0")
    },
  },
  {
    .bus = 1,
    .addr = 0x41,
    .channels = {
      INA3221_CHAN_CFG_INIT(1, "NC"),
      INA3221_CHAN_CFG_INIT(2, "VDDQ_VDD2_1V8AO"),
      INA3221_CHAN_CFG_INIT(3, "NC")
    },
  },
};

typedef struct jetson_dev_cfg {
  char name[JETSON_DEV_NAME_LEN];
  size_t n_ina3221_devs;
  ina3221_dev_cfg *ina3221_devs;
  // identifies by indexes which INA3221 devices and channels to open by default
  // TODO: might be easier to just add a flag array[3] to ina3221_dev_cfg? Does this break struct generality?
  int dev_chan_default_idxs[JETSON_INA3221_DEVS_MAX][3];
} jetson_dev_cfg;

static const jetson_dev_cfg JETSON_DEVS[] = {
  {
    .name = "Jetson Xavier NX Series",
    .n_ina3221_devs = 1,
    .ina3221_devs = DEVS_XAVIER_NX,
    .dev_chan_default_idxs = {{0},},
  },
  {
    .name = "Jetson AGX Xavier Series",
    .n_ina3221_devs = 2,
    .ina3221_devs = DEVS_AGX_XAVIER,
    .dev_chan_default_idxs = {{0, 1, 2}, {0, 1, 2}},
  },
  {
    .name = "Jetson AGX Orin Series",
    .n_ina3221_devs = 2,
    .ina3221_devs = DEVS_AGX_ORIN,
    // TODO: determine correct defaults
    .dev_chan_default_idxs = {{2}, {1}},
  },
};





static int init_libsensors(void) {
  const char* input;
  FILE* f = NULL;
  int rc;
  int err_save;
  if (getenv(ENERGYMON_JETSON_SENSORS_SKIP_LIFECYCLE)) {
    return 0;
  }
  if ((input = getenv(ENERGYMON_JETSON_SENSORS_INIT_FILE))) {
    if (!(f = fopen(input, "r"))) {
      fprintf(stderr, "fopen: %s: %s\n", input, strerror(errno));
      return -1;
    }
  }
  if ((rc = sensors_init(f))) {
    fprintf(stderr, "sensors_init: %s\n", sensors_strerror(rc));
    if (!errno) {
      // we don't really know the error, but we'll encourage user to retry
      errno = EAGAIN;
    }
  }
  if (f) {
    err_save = errno;
    if (fclose(f)) {
      fprintf(stderr, "fclose: %s: %s\n", input, strerror(errno));
    }
    errno = err_save;
  }
  return rc;
}

static void cleanup_libsensors(void) {
  if (!getenv(ENERGYMON_JETSON_SENSORS_SKIP_LIFECYCLE)) {
    sensors_cleanup();
  }
}





/**
 * Open all sensor files and start the thread to poll the sensors.
 */
int energymon_init_jetson_sensors(energymon* em) {
  if (em == NULL || em->state != NULL) {
    errno = EINVAL;
    return -1;
  }
  // TODO
  return -1;
}

int energymon_finish_jetson_sensors(energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return -1;
  }
  // TODO
  return -1;
}

uint64_t energymon_read_total_jetson_sensors(const energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return 0;
  }
  // TODO
  return 0;
}

char* energymon_get_source_jetson_sensors(char* buffer, size_t n) {
  return energymon_strencpy(buffer, "NVIDIA Jetson JetPack 5.x INA3221 Power Monitors", n);
}

uint64_t energymon_get_interval_jetson_sensors(const energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return 0;
  }
  // TODO
  return 0;
}

uint64_t energymon_get_precision_jetson_sensors(const energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return 0;
  }
  // milliwatts at refresh interval
  uint64_t prec = energymon_get_interval_jetson_sensors(em) / 1000;
  return prec ? prec : 1;
}

int energymon_is_exclusive_jetson_sensors(void) {
  return 0;
}

int energymon_get_jetson_sensors(energymon* em) {
  if (em == NULL) {
    errno = EINVAL;
    return -1;
  }
  em->finit = &energymon_init_jetson_sensors;
  em->fread = &energymon_read_total_jetson_sensors;
  em->ffinish = &energymon_finish_jetson_sensors;
  em->fsource = &energymon_get_source_jetson_sensors;
  em->finterval = &energymon_get_interval_jetson_sensors;
  em->fprecision = &energymon_get_precision_jetson_sensors;
  em->fexclusive = &energymon_is_exclusive_jetson_sensors;
  em->state = NULL;
  return 0;
}


jetson-sensors/energymon-jetson-sensors.h
/**
 * Energy reading for NVIDIA Jetson devices with TI INA3221 power sensors.
 *
 * @author Connor Imes
 * @date 2022-08-04
 */
#ifndef _ENERGYMON_JETSON_SENSORS_H_
#define _ENERGYMON_JETSON_SENSORS_H_

#ifdef __cplusplus
extern "C" {
#endif

#include <inttypes.h>
#include <stddef.h>
#include "energymon.h"

int energymon_init_jetson_sensors(energymon* em);

uint64_t energymon_read_total_jetson_sensors(const energymon* em);

int energymon_finish_jetson_sensors(energymon* em);

char* energymon_get_source_jetson_sensors(char* buffer, size_t n);

uint64_t energymon_get_interval_jetson_sensors(const energymon* em);

uint64_t energymon_get_precision_jetson_sensors(const energymon* em);

int energymon_is_exclusive_jetson_sensors(void);

int energymon_get_jetson_sensors(energymon* em);

#ifdef __cplusplus
}
#endif

#endif


raplcap-ipg/energymon-raplcap-ipg.c
/**
 * Read energy from Intel RAPL using the raplcap-msr library.
 *
 * @author Connor Imes
 * @date 2018-05-19
 */
#include <errno.h>
#include <inttypes.h>
#include <raplcap.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "energymon.h"
#include "energymon-raplcap-ipg.h"
#include "energymon-util.h"

#ifdef ENERGYMON_DEFAULT
#include "energymon-default.h"
int energymon_get_default(energymon* em) {
  return energymon_get_raplcap_ipg(em);
}
#endif

typedef struct raplcap_ipg_info {
  double j_last;
  double j_max;
  uint32_t n_overflow;
} raplcap_ipg_info;

typedef struct energymon_raplcap_ipg {
  raplcap rc;
  uint32_t n_pkg;
  uint32_t n_die;
  uint32_t n_msrs;
  raplcap_ipg_info msrs[];
} energymon_raplcap_ipg;

int energymon_init_raplcap_ipg(energymon* em) {
  if (em == NULL || em->state != NULL) {
    errno = EINVAL;
    return -1;
  }

  int err_save;
  int supp;
  uint32_t i;
  uint32_t n_pkg;
  uint32_t n_die;
  uint32_t n_msrs;
  uint32_t max_die;
  uint32_t pkg;
  uint32_t die;
  if ((n_pkg = raplcap_get_num_packages(NULL)) == 0) {
    perror("raplcap_get_num_packages");
    return -1;
  }
  // currently only support systems with homogeneous package/die configurations
  for (pkg = 0; pkg < n_pkg; pkg++) {
    if ((n_die = raplcap_get_num_die(NULL, pkg)) == 0) {
      if (errno == ENOSYS) {
        // TODO: for now, assume single-die
        n_die = 1;
      } else {
        perror("raplcap_get_num_die");
        return -1;
      }
    }
    if (pkg == 0) {
      max_die = n_die;
    } else if (max_die != n_die) {
      fprintf(stderr, "energymon_init_raplcap_ipg: no support for heterogeneous package/die\n");
      errno = ENOSYS;
      return -1;
    }
  }
  n_msrs = n_pkg * n_die;

  energymon_raplcap_ipg* state = calloc(1, sizeof(energymon_raplcap_ipg) + (n_msrs * sizeof(raplcap_ipg_info)));
  if (state == NULL) {
    return -1;
  }
  state->n_msrs = n_msrs;
  state->n_pkg = n_pkg;
  state->n_die = n_die;

  if (raplcap_init(&state->rc)) {
    perror("raplcap_init");
    free(state);
    return -1;
  }

  for (pkg = 0; pkg < state->n_pkg; pkg++) {
    for (die = 0; die < state->n_die; die++) {
      i = pkg * die + die;
      // first check if zone is supported
      supp = raplcap_pd_is_zone_supported(&state->rc, pkg, die, RAPLCAP_ZONE_PACKAGE);
      if (supp < 0) {
        perror("raplcap_pd_is_zone_supported");
        goto fail;
      }
      if (supp == 0) {
        fprintf(stderr, "energymon_init_raplcap_ipg: Unsupported zone: PACKAGE\n");
        errno = EINVAL;
        goto fail;
      }
      // Note: max energy is specified in a different MSR than the zone's energy counter,
      // so this call might still work for unsupported zones (which is why we have to check for support first)
      if ((state->msrs[i].j_max = raplcap_pd_get_energy_counter_max(&state->rc, pkg, die, RAPLCAP_ZONE_PACKAGE)) < 0) {
        if (errno == ENOSYS) {
          // TODO: this is just for testing
          state->msrs[i].j_max = 1000000000;
        } else {
          perror("raplcap_pd_get_energy_counter_max");
          goto fail;
        }
      }
    }
  }

  em->state = state;
  return 0;

fail:
  err_save = errno;
  if (raplcap_destroy(&state->rc)) {
    perror("raplcap_destroy");
  }
  free(state);
  errno = err_save;
  return -1;
}

uint64_t energymon_read_total_raplcap_ipg(const energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return 0;
  }
  uint32_t pkg;
  uint32_t die;
  uint32_t i;
  double j;
  uint64_t total = 0;
  energymon_raplcap_ipg* state = (energymon_raplcap_ipg*) em->state;
  for (errno = 0, pkg = 0; pkg < state->n_pkg && !errno; pkg++) {
    for (die = 0; die < state->n_die && !errno; die++) {
      i = pkg * die + die;
      if ((j = raplcap_pd_get_energy_counter(&state->rc, pkg, die, RAPLCAP_ZONE_PACKAGE)) >= 0) {
        if (j < state->msrs[i].j_last) {
          state->msrs[i].n_overflow++;
        }
        state->msrs[i].j_last = j;
        total += (uint64_t) ((j + state->msrs[i].n_overflow * state->msrs[i].j_max) * 1000000.0);
      }
    }
  }
  return errno ? 0 : total;
}

int energymon_finish_raplcap_ipg(energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return -1;
  }
  int ret;
  energymon_raplcap_ipg* state = em->state;
  if ((ret = raplcap_destroy(&state->rc))) {
    perror("raplcap_destroy");
  }
  free(em->state);
  em->state = NULL;
  return ret;
}

char* energymon_get_source_raplcap_ipg(char* buffer, size_t n) {
  return energymon_strencpy(buffer, "Intel RAPL via libraplcap-ipg", n);
}

uint64_t energymon_get_interval_raplcap_ipg(const energymon* em) {
  if (em == NULL) {
    // we don't need to access em, but it's still a programming error
    errno = EINVAL;
    return 0;
  }
  return 1000;
}

uint64_t energymon_get_precision_raplcap_ipg(const energymon* em) {
  if (em == NULL || em->state == NULL) {
    errno = EINVAL;
    return 0;
  }
  // TODO: can we determine this from raplcap-ipg?
  return 1;
}

int energymon_is_exclusive_raplcap_ipg(void) {
  return 0;
}

int energymon_get_raplcap_ipg(energymon* em) {
  if (em == NULL) {
    errno = EINVAL;
    return -1;
  }
  em->finit = &energymon_init_raplcap_ipg;
  em->fread = &energymon_read_total_raplcap_ipg;
  em->ffinish = &energymon_finish_raplcap_ipg;
  em->fsource = &energymon_get_source_raplcap_ipg;
  em->finterval = &energymon_get_interval_raplcap_ipg;
  em->fprecision = &energymon_get_precision_raplcap_ipg;
  em->fexclusive = &energymon_is_exclusive_raplcap_ipg;
  em->state = NULL;
  return 0;
}
